core/src/main/kotlin/org/jetbrains/research/testspark/core/generation/llm/LLMWithFeedbackCycle.kt

```diff
@@ -60,226 +44,146 @@ class LLMWithFeedbackCycle(
+
+                val chunks: Flow<Result<String>> =
+                    chatSessionManager.request(
+                        prompt = nextPromptMessage,
+                        isUserFeedback = false,
+                    )
+                val testSuiteResult: Result<TestSuiteGeneratedByLLM> = chunks.collectChunks(testsAssembler)
+
+                when (testSuiteResult) {
+                    is Result.Success -> log.info { "Test suite generated successfully: ${testSuiteResult.data}" }
+
+                    is Result.Failure -> {
+                        log.info { "Cannot parse a test suite from the LLM response. LLM response: '$testSuiteResult'" }
+                        emit(testSuiteResult)
+                        nextPromptMessage = generatePromptMessage(testSuiteResult.error) ?: break
+
+                        /**
+                         * The current attempt does not count as a failure since it was rejected due to the prompt size
+                         * exceeding the threshold
+                         */
+                        if (testSuiteResult.error is LlmError.PromptTooLong) iteration--
                         continue
                     }
                 }
 
-                is Result.Failure -> {
-                    when (response.error) {
-                        is LlmError.EmptyLlmResponse -> {
-                            nextPromptMessage =
-                                "You have provided an empty answer! Please, answer my previous question with the same formats"
-                            continue
-                        }
-
-                        is LlmError.PromptTooLong -> {
-                            if (promptSizeReductionStrategy.isReductionPossible()) {
-                                nextPromptMessage = promptSizeReductionStrategy.reduceSizeAndGeneratePrompt()
-                                /**
-                                 * The current attempt does not count as a failure
-                                 * since it was rejected due to the prompt size
-                                 * exceeding the threshold
-                                 */
-                                requestsCount--
-                                continue
-                            } else {
-                                return Result.Failure(error = LlmError.PromptTooLong)
-                            }
-                        }
-
-                        is LlmError.TestSuiteParsingError -> {
-                            onWarningCallback?.invoke(LlmError.TestSuiteParsingError)
-                            log.info { "Cannot parse a test suite from the LLM response. LLM response: '$response'" }
-
-                            nextPromptMessage = "The provided code is not parsable. Please, generate the correct code"
-                            continue
-                        }
+                val testSuite = testSuiteResult.data
+                generatedTestSuites.add(testSuite)
```

Comment:
generatedTestSuites is not cleared between two cycles. At the end of the process, all generated tests from every iteration are displayed. 
In my opinion, we only need to keep compilable ones or only from the latest cycle.